apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,327 @@
+delve into how we extended the support for `torch.func` transforms. Given that
+`aot_autograd` is already capable of tracing through the transforms, our task is
+to teach `dynamo` to validate whether the user-defined function intended for
+transformation is free of side effects affecting the global state or graph-breaks.
+In cases where the function meets these criteria, we can put
+the `torch.func` transform into the FX graph and delegate the remaining
+processing to the lower layers of the stack.
+
+However, if the function cannot be successfully traced due to its failure to
+meet the above constraints, we fallback to the eager implementation, and this
+particular portion of the code remains uncompiled.
+
+Let us have a look what `dynamo` and `aot_autograd` generates when we compile
+program with `grad`.
+
+```python
+# Run this file with `TORCH_COMPILE_DEBUG=1`
+
+import torch
+
+def user_fn(x):
+    return torch.sin(x)
+
+def wrapper_fn(x):
+    return torch.func.grad(user_fn)(x)
+
+torch.compile(wrapper_fn)(torch.randn(()))
+
+```
+
+The output from `dynamo` is presented below. The initial `GraphModule` pertains
+to the `wrapper_fn`, clearly indicating a call to `grad` on the traced representation
+of the user's function intended for transformation. Subsequently, the second
+`GraphModule` corresponds to the function provided by the user. In this instance,
+our function didn't have side effects and graph-breaks. Thus, we were able to
+successfully trace through this program in one graph.
+
+```python
+class GraphModule(torch.nn.Module):
+    def forward(self, L_x_ : torch.Tensor):
+        l_x_ = L_x_
+        
+        # File: torch/_functorch/apis.py:363, code:
+        # return eager_transforms.grad_impl(func, argnums, has_aux, args, kwargs)
+        grad_body_0 = self.grad_body_0
+        grad_proxy = torch.func.grad(grad_body_0, 0, False);  grad_body_0 = None
+        call = grad_proxy.__call__(l_x_);  grad_proxy = l_x_ = None
+        contiguous = call.contiguous();  call = None
+        return (contiguous,)
+        
+    class GraphModule(torch.nn.Module):
+        def forward(self, l_x_):
+            # No stacktrace found for following nodes
+            _set_grad_enabled = torch._C._set_grad_enabled(True)
+
+            # File: test/test_scratch.py:382, code: return torch.sin(x)
+            sin = torch.sin(l_x_);  l_x_ = None
+
+            # No stacktrace found for following nodes
+            _set_grad_enabled_1 = torch._C._set_grad_enabled(True)
+            return sin
+```
+
+The graph shown above is handed over to `aot_autograd` for the subsequent phase
+of the compilation process. `aot_autograd` performs a trace through the
+transformation, resulting in the generation of the transformed graph. This
+explains why we observe a call to `cos` instead of `sin`, as the transformation
+has effectively altered the operation from what was originally present in our function.
```

Comment:
We see `cos` because it's the backward of `sin`, right? It would be nice to highlight that aot_autograd has traced both forward and backward and simplified the graph to just the backward computation.

Reply:
How does this patch look?

```patch
-explains why we observe a call to `cos` instead of `sin`, as the transformation
-has effectively altered the operation from what was originally present in our function.
+explains why we observe a call to `cos` instead of `sin`. `aot_autograd`
+has traced through the forward and backward graph as we have applied `grad` transform
+and optimized away the forward computation as `grad` discards that value.
```